src/ngp_algorithms/NodalGradPOpenBoundaryAlg.C

```diff
@@ -0,0 +1,161 @@
+}
+
+
+
+
+//--------------------------------------------------------------------------
+//-------- execute ---------------------------------------------------------
+//--------------------------------------------------------------------------
+template <typename AlgTraits>
+void
+NodalGradPOpenBoundary<AlgTraits>::execute()
+{
+  using SimdDataType = nalu_ngp::FaceElemSimdData<ngp::Mesh>;
+
+  const auto& meshInfo = realm_.mesh_info();
+  const auto& meta_data = meshInfo.meta();
+
+  const bool useShifted = useShifted_;
+  const bool zeroGrad   = zeroGrad_;
+
+  const unsigned exposedAreaVec       = exposedAreaVec_;
+  const unsigned dualNodalVol         = dualNodalVol_;
+  const unsigned exposedPressureField = exposedPressureField_;
+  const unsigned pressureField        = pressureField_;
+  const unsigned coordsID             = coordinates_;
+
+  const auto& fieldMgr = meshInfo.ngp_field_manager();
+  const auto   ngpMesh = meshInfo.ngp_mesh();
+  auto  gradP = fieldMgr.template get_field<double>(gradP_);
+  const auto gradPOps = nalu_ngp::simd_face_elem_nodal_field_updater(ngpMesh, gradP);
+
+  MasterElement* meFC  = meFC_;
+  MasterElement* meSCS = meSCS_;
+
+  stk::mesh::Selector s_locally_owned_union = meta_data.locally_owned_part() & stk::mesh::selectUnion(partVec_);
+
+  const std::string algName = "NodalGradPOpenBoundary_" +
+                              std::to_string(AlgTraits::faceTopo_) + "_" +
+                              std::to_string(AlgTraits::elemTopo_);
+
+  nalu_ngp::run_face_elem_algorithm(
+    algName, meshInfo, faceData_, elemData_, s_locally_owned_union,
+    KOKKOS_LAMBDA(SimdDataType& fdata) {
+
+      const int* ipNodeMap = meFC->ipNodeMap();
+
+      auto& faceView = fdata.simdFaceView;
+      const auto v_areav        = faceView.get_scratch_view_2D(exposedAreaVec);
+      const auto v_dnv          = faceView.get_scratch_view_1D(dualNodalVol);
+      const auto face_p_field   = faceView.get_scratch_view_1D(exposedPressureField);
+      const auto elem_p_field   = faceView.get_scratch_view_1D(pressureField);
+      const auto v_coord        = faceView.get_scratch_view_2D(coordsID);
+
+      const auto meFaceViews = fdata.simdFaceView.get_me_views(CURRENT_COORDINATES);
+      const auto meElemViews = fdata.simdElemView.get_me_views(CURRENT_COORDINATES);
+      const auto v_shape_fcn = useShifted ? meFaceViews.fc_shifted_shape_fcn :  meFaceViews.fc_shape_fcn;
+      const auto e_shape_fcn = useShifted ? meElemViews.scs_shifted_shape_fcn:  meElemViews.scs_shape_fcn;
+
+      const int faceOrdinal = fdata.faceOrd;
+
+      for (int ip = 0; ip < AlgTraits::numFaceIp_; ++ip) {
+        DoubleType pIp = 0.0;
+        if (zeroGrad) {
+          // evaluate pressure at opposing face.  
+          const int oip = meSCS->opposingFace(faceOrdinal, ip);
+          for (int n = 0; n < AlgTraits::nodesPerElement_; ++n) {
+            pIp += e_shape_fcn(oip,n) * elem_p_field(n);
```

Comment:
@overfelt Sorry, I should have looked closer. This is still a bug, you are accessing `elem_p_field` as a HEX field, but it is registered as a face (QUAD_4) field (See line 65).

Reply:
Shreyas,  Yes, this is a bug.  The pressure field should have been registered as an element gathered field, not a face gathered field.